### PR TITLE
Use unique list of address to create snapshots instead of auth module

### PR DESCRIPTION
**Pull request by politeWall**

### CODE REVIEW

**Comment by politeWall** [comment]
## Test creating snapshot

### query snapshot
```sh
elysd q --output json --node "tcp://localhost:26657" wasm contract-state smart "elys17p9rzwnnfxcjp32un9ug7yhhzgtkhvl9jfksztgw5uh69wac2pgs98tvuy" '{
        "user_snapshots": {
            "user_address": "elys1pszy7cq5xhx6dvu082nvyjjd2q6mln9n3k5ly0"
        }
    }' | jq
```
```json
{
  "data": []
}
```
### using the `trade-shield-contract`

```sh
elysd tx wasm exec --from "validator" --keyring-backend test --node "tcp://localhost:26657" --chain-id elystestnet-1 --gas auto --gas-adjustment=1.3 --fees 100000uelys -b sync -y --amount 1000000uelys "elys1nc5tatafv6eyq7llkr2gv50ff9e22mnf70qgjlv737ktmt4eswrqau4f4q" '{
            "amm_swap_exact_amount_in": {
                "routes": [
                    {
                        "pool_id": 3,
                        "token_out_denom": "ibc/2180E84E20F5679FCC760D8C165B60F42065DEF7F46A72B447CFF1B7DC6C0A65"
                    }
                ]
            }
        }'
```
### query snapshot
```json
{
  "data": [
    {
      "date": {
        "at_time": "1714492381690649093"
      },
      "portfolio_balance_usd": "4775676449.41425133102773063",
      "total_balance_usd": "4775676449.41425133102773063"
    }
  ]
}
```
### using the `trade-shield-contract`


### query snapshot
no new snapshot and still have the old snapshot of the day
```json
{
  "data": [
    {
      "date": {
        "at_time": "1714492381690649093"
      },
      "portfolio_balance_usd": "4775676449.41425133102773063",
      "total_balance_usd": "4775676449.41425133102773063"
    }
  ]
}
```

**Comment by politeWall** [comment]
### Get Portfolio

if there is no snapshot available in the past 24hrs for a given account, then the 24hrs balance change is showing 0 instead of the full balance amount.
```json
{
  "data": {
    "portfolio": {
      "balance_usd": "4775555633.533259094595382036",
      "liquid_assets_usd": "4775555633.533259094595382036",
      "staked_committed_usd": "0",
      "liquidity_positions_usd": "0",
      "leverage_lp_usd": "0",
      "perpetual_assets_usd": "0",
      "usdc_earn_usd": "0",
      "borrows_usd": "0",
      "stake_balance_breakdown": {
        "staked": "0",
        "unstaking": "0",
        "vesting": "0"
      }
    },
    "actual_portfolio_balance": "4775555633.533259094595382036",
    "old_portfolio_balance": "0",
    "balance_24h_change": "0"
  }
}
```

**Comment by politeWall** [comment]
### GetMembershipTier

```json
{
  "data": {
    "identifier": "platinum",
    "name": "Platinum",
    "discount": "0.3"
  }
}
```

**Comment by politeWall** [comment]
### Get Portfolio 24h change
```json
{
  "data": {
    "portfolio": {
      "balance_usd": "4775555633.533259094595382036",
      "liquid_assets_usd": "4775555633.533259094595382036",
      "staked_committed_usd": "0",
      "liquidity_positions_usd": "0",
      "leverage_lp_usd": "0",
      "perpetual_assets_usd": "0",
      "usdc_earn_usd": "0",
      "borrows_usd": "0",
      "stake_balance_breakdown": {
        "staked": "0",
        "unstaking": "0",
        "vesting": "0"
      }
    },
    "actual_portfolio_balance": "4775555633.533259094595382036",
    "old_portfolio_balance": "4775676449.41425133102773063",
    "balance_24h_change": "-120815.880992236432348594"
  }
}
```

File: contracts/account-history-contract/src/action/query/user_snapshots.rs [hunk 1]
```diff
@@ -1,24 +1,48 @@
...
+
+        let day_history = HISTORY.may_load(deps.storage, &date)?;
+        if let Some(day_history) = day_history {
+            if let Some(portfolio) = day_history.get(&user_address) {
+                user_snapshots_list.push(portfolio.to_owned())
+            };
+        }
+
+        day_date = day_date.plus_days(1);
+    }
```
Comment (cosmic-vagabond) [inline contracts/account-history-contract/src/action/query/user_snapshots.rs:45]: @politeWall this code is the same across multiple functions, can you move it to a sub function?

File: contracts/account-history-contract/src/action/sudo/update_account.rs [hunk 2]
```diff
@@ -20,129 +19,110 @@ pub fn update_account(deps: DepsMut<ElysQuery>, env: Env) -> StdResult<Response<
...
-        let snapshot = PortfolioBalanceSnapshot {
-            date: Expiration::AtTime(Timestamp::from_seconds(1707306681)),
-            total_balance_usd: Decimal256::zero(),
-            portfolio_balance_usd: Decimal256::zero(),
-        };
+        // let snapshot = PortfolioBalanceSnapshot {
+        //     date: Expiration::AtTime(Timestamp::from_seconds(1707306681)),
+        //     total_balance_usd: Decimal256::zero(),
+        //     portfolio_balance_usd: Decimal256::zero(),
+        // };
```
Comment (cosmic-vagabond) [inline contracts/account-history-contract/src/action/sudo/update_account.rs:102]: @politeWall can we get rid of those commented lines?

File: contracts/account-history-contract/src/action/sudo/update_account.rs [hunk 3]
```diff
@@ -20,129 +19,110 @@ pub fn update_account(deps: DepsMut<ElysQuery>, env: Env) -> StdResult<Response<
...
-            "failed to get accounts with pagination {:?}: {}",
-            pagination, e
-        ))
-    })?;
-
-    pagination.update(resp.pagination.next_key);
-    PAGINATION.save(deps.storage, &pagination)?;
-
     let today = get_today(&env.block);
+    let user_address_queue: Vec<String> = USER_ADDRESS_QUEUE
```
Comment (cosmic-vagabond) [inline contracts/account-history-contract/src/action/sudo/update_account.rs:23]: @politeWall can you also add a new query that will provide the current number of user addresses in the queue, we want to make sure that we process enough addresses per block (limit set by default to 1) to make up for all the addresses in a 24 hrs window

**Review by cosmic-vagabond — CHANGES_REQUESTED** [review]

File: contracts/account-history-contract/src/action/query/user_snapshots.rs (hunk 1 above)
Comment (politeWall) [reply contracts/account-history-contract/src/action/query/user_snapshots.rs:45]: Yes I use the function [user_snapshots](https://github.com/elys-network/bindings/pull/301/files#diff-dff06becce0fece20d324cb5f95b53ca32f22439b9901d3b00de373b999eeb57) to replace the duplicate code

File: contracts/account-history-contract/src/action/sudo/update_account.rs (hunk 2 above)
Comment (politeWall) [reply contracts/account-history-contract/src/action/sudo/update_account.rs]: yes done

File: contracts/account-history-contract/src/action/sudo/update_account.rs (hunk 3 above)
Comment (politeWall) [reply contracts/account-history-contract/src/action/sudo/update_account.rs:23]: yes query added [here](https://github.com/elys-network/bindings/pull/301/files#diff-fc69f4d4e3b05fb0eb6e7618fa7c1e13f6b8d8b4f19efdc7d93debe31a2cc8f9R166-R175)

**Pull request merged**